tests/kafkatest/tests/client/downconversion_memory_test.py

```diff
@@ -0,0 +1,119 @@
+# The ASF licenses this file to You under the Apache License, Version 2.0
+# (the "License"); you may not use this file except in compliance with
+# the License.  You may obtain a copy of the License at
+#
+#    http://www.apache.org/licenses/LICENSE-2.0
+#
+# Unless required by applicable law or agreed to in writing, software
+# distributed under the License is distributed on an "AS IS" BASIS,
+# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+# See the License for the specific language governing permissions and
+# limitations under the License.
+from ducktape.mark import parametrize
+from ducktape.mark.resource import cluster
+from ducktape.tests.test import Test
+
+from kafkatest.services.kafka import KafkaService
+from kafkatest.services.performance import ProducerPerformanceService, compute_aggregate_throughput, \
+    ConsumerPerformanceService
+from kafkatest.services.zookeeper import ZookeeperService
+from kafkatest.version import DEV_BRANCH, LATEST_0_10, KafkaVersion, LATEST_1_1, LATEST_0_10_0
+
+
+class DownconversionMemoryTest(Test):
+    def __init__(self, test_context):
+        super(DownconversionMemoryTest, self).__init__(test_context=test_context)
+        '''
+        Test Setup:
+        ==========
+        - Java heap size = 190MB
+        - 1M messages, 1kB each ==> 1GB of total messages
+        - Split into 200 partitions ==> approximately 5MB per partition
+        - 1 consumer with `fetch.max.bytes` = 200MB and `max.partition.fetch.bytes` = 1MB
+        - Each fetch consumes min(1MB*200, 200MB) = 200MB i.e. 1MB from each partition for a total of 200MB
+        - Success criteria:
+            - Must always run out of memory if not using lazy down-conversion
+            - Must never run out of memory if using lazy down-conversion
+        '''
+        self.heap_size = 190
+        self.max_messages = 1024 * 1024
+        self.message_size = 1024
+        self.batch_size = self.message_size * 50
+        self.num_partitions = 200
+        self.max_fetch_size = 200 * 1024 * 1024
+        self.num_producers = 1
+        self.num_consumers = 1
+        self.topics = ["test_topic"]
+        self.zk = ZookeeperService(self.test_context, num_nodes=1)
+        self.heap_dump_path = "/mnt/heapdump/"
+        self.consumer_version = LATEST_0_10
+
+    def setUp(self):
+        self.zk.start()
+        super(DownconversionMemoryTest, self).setUp()
+
+    def tearDown(self):
+        for node in self.kafka.nodes:
+            node.account.ssh("rm -rf %s" % self.heap_dump_path, allow_fail=False)
+        super(DownconversionMemoryTest, self).tearDown()
+
+    @cluster(num_nodes=12)
+    @parametrize(version=str(LATEST_1_1))
+    @parametrize(version=str(DEV_BRANCH))
+    def test_downconversion(self, version):
+        broker_version = KafkaVersion(version)
+        producer_version = broker_version
+
+        out_of_memory = False
+        if KafkaVersion(version) <= LATEST_1_1:
```

Comment:
Since we're not actually testing this case, do you think it's still worthwhile to have these checks?

Reply:
I am not sure I follow, we are testing with `LATEST_1_1` as one of the parameterizations for this test. This is more of a validation for the test setup itself - what we want to make sure is the test setup is such that we run out of memory on older versions of the broker.